src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -956,6 +1076,64 @@ public DocumentId AddTextContainer(SourceTextContainer textContainer, string ful
                 return documentId;
             }
 
+            public void AddDynamicFile(IDynamicFileInfoProvider fileInfoProvider, DynamicFileInfo fileInfo, ImmutableArray<string> folders)
+            {
+                var documentInfo = CreateDocumentInfoFromFileInfo(fileInfo, folders.IsDefault ? null : (IEnumerable<string>)folders);
+                var documentId = documentInfo.Id;
+
+                lock (_project._gate)
+                {
+                    var filePath = documentInfo.FilePath;
+                    if (_documentPathsToDocumentIds.ContainsKey(filePath))
+                    {
+                        throw new ArgumentException($"'{filePath}' has already been added to this project.", nameof(filePath));
+                    }
+
+                    _documentPathsToDocumentIds.Add(filePath, documentId);
+
+                    _documentIdToDynamicFileInfoProvider.Add(documentId, fileInfoProvider);
+
+                    if (_project._eventSubscriptionTracker.Add(fileInfoProvider))
+                    {
+                        // subscribe to the event when we use this provider the first time
+                        fileInfoProvider.Updated += _project.OnDynamicFileInfoUpdated;
+                    }
+
+                    if (_project._activeBatchScopes > 0)
+                    {
+                        _documentsAddedInBatch.Add(documentInfo);
+                    }
+                    else
+                    {
+                        // right now, assumption is dynamically generated file can never be opened in editor
+                        _project._workspace.ApplyChangeToWorkspace(w => _documentAddAction(w, documentInfo));
+                    }
+                }
+            }
+
+            public IDynamicFileInfoProvider RemoveDynamicFile(string fullPath)
+            {
+                if (string.IsNullOrEmpty(fullPath))
+                {
+                    throw new ArgumentException($"{nameof(fullPath)} isn't a valid path.", nameof(fullPath));
+                }
+
+                lock (_project._gate)
+                {
+                    if (!_documentPathsToDocumentIds.TryGetValue(fullPath, out var documentId) ||
+                        !_documentIdToDynamicFileInfoProvider.TryGetValue(documentId, out var fileInfoProvider))
+                    {
+                        throw new ArgumentException($"'{fullPath}' is not a dynamic file of this project.");
+                    }
+
+                    _documentIdToDynamicFileInfoProvider.Remove(documentId);
```

Comment:
feels super weird this needs to be done.  it feels like we hav emultiple sources of truth here.

Reply:
at the very least, if we do need all these maps, it feel slike they should be consolidated into one type that appropriate protects, mutates, and allows querying of hte data.

Reply:
could be. but not in this PR. I dont think it is different than anything already exist in this code.

Reply:
Could you explain why?  If we're introducing all this new code, we should ideally be doing so in a healthy and well-factored manner.  Just splatting all this stuff all over the place makes things much more complex and hard to reason about.  Good factorization helps ensure that invariants can clearly be documented and enforced (something lacking in this PR), as well as makign it clear what the expected surface area is for others that need to consume this new functionality (also not clear in this PR).  

We have many language mechanisms at our disposal with which we can write code with varying levels of quality.  Non-onerous patterns and practices should be adopted by default to keep our codebase as healthy as possible.